File: src/modules/onedrive_backup.py
"""
OneDrive备份模块 - 使用Microsoft Graph API备份已会词到OneDrive

功能：
1. OAuth2授权流程（设备码流，无需公网回调）
2. 全量备份已会词到OneDrive（JSON格式，带版本号）
3. 列出备份历史
4. 从OneDrive恢复（合并到本地数据库）
"""

import json
import os
import time
import sqlite3
from pathlib import Path
from typing import Optional, List, Dict, Any
from datetime import datetime

import requests

TOKEN_FILE = "onedrive_token.json"
BACKUP_FOLDER = "Apps/DreamWord"

GRAPH_BASE = "https://graph.microsoft.com/v1.0"


class OneDriveBackup:
    def __init__(self, client_id: str, db_path: str = "known_words.db"):
        self.client_id = client_id
        self.db_path = db_path
        self.token_path = TOKEN_FILE
        self._tokens = None

    def _load_tokens(self) -> Optional[Dict[str, Any]]:
        if self._tokens is not None:
            return self._tokens
        if not os.path.exists(self.token_path):
            return None
        try:
            with open(self.token_path, 'r', encoding='utf-8') as f:
                self._tokens = json.load(f)
            return self._tokens
        except Exception:
            return None

    def _save_tokens(self, tokens: Dict[str, Any]):
        self._tokens = tokens
        with open(self.token_path, 'w', encoding='utf-8') as f:
            json.dump(tokens, f, indent=2, ensure_ascii=False)

    def get_device_code(self) -> Dict[str, Any]:
        resp = requests.post(
            "https://login.microsoftonline.com/common/oauth2/v2.0/devicecode",
            data={
                "client_id": self.client_id,
                "scope": "Files.ReadWrite.AppFolder offline_access"
            },
            timeout=30
        )
        if resp.status_code != 200:
            raise Exception(f"获取设备码失败: {resp.text}")
        return resp.json()

    def poll_token(self, device_code: str) -> Optional[Dict[str, Any]]:
        resp = requests.post(
            "https://login.microsoftonline.com/common/oauth2/v2.0/token",
            data={
                "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
                "client_id": self.client_id,
                "device_code": device_code
            },
            timeout=30
        )
        if resp.status_code == 200:
            tokens = resp.json()
            tokens['expires_at'] = time.time() + tokens.get('expires_in', 3600) - 60
            self._save_tokens(tokens)
            return tokens
        error = resp.json().get('error', '')
        if error == 'authorization_pending':
            return None
        if error == 'slow_down':
            return None
        raise Exception(f"获取token失败: {resp.text}")

    def _ensure_token(self) -> str:
        tokens = self._load_tokens()
        if not tokens:
            raise Exception("未授权，请先完成OneDrive授权")
        if time.time() > tokens.get('expires_at', 0):
            self._refresh_token()
            tokens = self._load_tokens()
        return tokens['access_token']

    def _refresh_token(self):
        tokens = self._load_tokens()
        if not tokens or 'refresh_token' not in tokens:
            raise Exception("无refresh_token，请重新授权")
        resp = requests.post(
            "https://login.microsoftonline.com/common/oauth2/v2.0/token",
            data={
                "grant_type": "refresh_token",
                "client_id": self.client_id,
                "refresh_token": tokens['refresh_token']
            },
            timeout=30
        )
        if resp.status_code != 200:
            raise Exception(f"刷新token失败: {resp.text}")
        new_tokens = resp.json()
        new_tokens['expires_at'] = time.time() + new_tokens.get('expires_in', 3600) - 60
        self._save_tokens(new_tokens)
        self._tokens = new_tokens

    def _graph_get(self, path: str) -> Any:
        token = self._ensure_token()
        resp = requests.get(
            f"{GRAPH_BASE}{path}",
            headers={"Authorization": f"Bearer {token}"},
            timeout=30
        )
        if resp.status_code == 401:
            self._refresh_token()
            token = self._ensure_token()
            resp = requests.get(
                f"{GRAPH_BASE}{path}",
                headers={"Authorization": f"Bearer {token}"},
                timeout=30
            )
        if resp.status_code != 200:
            raise Exception(f"Graph API错误: {resp.status_code} {resp.text}")
        return resp.json()

    def _graph_put(self, path: str, content: bytes) -> Any:
        token = self._ensure_token()
        resp = requests.put(
            f"{GRAPH_BASE}{path}",
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json"
            },
            data=content,
            timeout=60
        )
        if resp.status_code == 401:
            self._refresh_token()
            token = self._ensure_token()
            resp = requests.put(
                f"{GRAPH_BASE}{path}",
                headers={
                    "Authorization": f"Bearer {token}",
                    "Content-Type": "application/json"
                },
                data=content,
                timeout=60
            )
        if resp.status_code not in (200, 201):
            raise Exception(f"上传失败: {resp.status_code} {resp.text}")
        return resp.json()

    def _graph_patch(self, path: str, content: bytes) -> Any:
        token = self._ensure_token()
        resp = requests.patch(
            f"{GRAPH_BASE}{path}",
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json"
            },
            data=content,
            timeout=30
        )
        if resp.status_code not in (200, 201):
            raise Exception(f"更新失败: {resp.status_code} {resp.text}")
        return resp.json()

    def _graph_post(self, path: str, content: bytes) -> Any:
        token = self._ensure_token()
        resp = requests.post(
            f"{GRAPH_BASE}{path}",
            headers={
                "Authorization": f"Bearer {token}",
                "Content-Type": "application/json"
            },
            data=content,
            timeout=30
        )
        if resp.status_code not in (200, 201):
            raise Exception(f"请求失败: {resp.status_code} {resp.text}")
        return resp.json()

    def is_authorized(self) -> bool:
        tokens = self._load_tokens()
        if not tokens:
            return False
        try:
            self._ensure_token()
            return True
        except Exception:
            return False

    def _get_all_words(self) -> List[str]:
        if not os.path.exists(self.db_path):
            return []
        conn = sqlite3.connect(self.db_path)
        try:
            cursor = conn.cursor()
            cursor.execute('SELECT word FROM known_words ORDER BY word')
            return [row[0] for row in cursor.fetchall()]
        finally:
            conn.close()

    def _get_local_version(self) -> int:
        meta_path = Path(self.db_path).parent / "backup_meta.json"
        if meta_path.exists():
            try:
                with open(meta_path, 'r', encoding='utf-8') as f:
                    return json.load(f).get('last_backup_version', 0)
            except Exception:
                return 0
        return 0

    def _set_local_version(self, version: int):
        meta_path = Path(self.db_path).parent / "backup_meta.json"
        with open(meta_path, 'w', encoding='utf-8') as f:
            json.dump({'last_backup_version': version}, f)

    def _ensure_folder(self):
        try:
            self._graph_get("/me/drive/special/approot/children")
        except Exception:
            pass

    def backup(self) -> Dict[str, Any]:
        words = self._get_all_words()
        local_version = self._get_local_version()
        new_version = local_version + 1

        backup_data = {
            "version": new_version,
            "timestamp": datetime.now().isoformat(),
            "word_count": len(words),
            "words": words
        }

        self._ensure_folder()

        filename = f"known_words_v{new_version}.json"
        content = json.dumps(backup_data, ensure_ascii=False, indent=2).encode('utf-8')

        self._graph_put(
            f"/me/drive/special/approot:/{filename}:/content",
            content
        )

        try:
            meta_content = json.dumps({
                "last_backup_version": new_version,
                "last_backup_time": backup_data["timestamp"],
                "last_backup_count": len(words)
            }).encode('utf-8')
            self._graph_put(
                "/me/drive/special/approot:/backup_meta.json:/content",
                meta_content
            )
        except Exception:
            pass

        self._set_local_version(new_version)

        return {
            "success": True,
            "version": new_version,
            "word_count": len(words),
            "timestamp": backup_data["timestamp"]
        }

    def list_backups(self) -> List[Dict[str, Any]]:
        try:
            result = self._graph_get("/me/drive/special/approot/children")
        except Exception:
            return []

        backups = []
        for item in result.get('value', []):
            name = item.get('name', '')
            if name.startswith('known_words_v') and name.endswith('.json'):
                backups.append({
                    'name': name,
                    'id': item.get('id', ''),
                    'size': item.get('size', 0),
                    'last_modified': item.get('lastModifiedDateTime', ''),
                    'download_url': item.get('@content.downloadUrl', '')
                })
        backups.sort(key=lambda x: x['name'], reverse=True)
        return backups

    def restore(self, backup_name: Optional[str] = None, merge: bool = True) -> Dict[str, Any]:
        if backup_name:
            path = f"/me/drive/special/approot:/{backup_name}:/content"
        else:
            meta = self._graph_get("/me/drive/special/approot:/backup_meta.json:/content")
            path = f"/me/drive/special/approot:/known_words_v{meta['last_backup_version']}.json:/content"

        token = self._ensure_token()
        resp = requests.get(
            f"{GRAPH_BASE}{path}",
            headers={"Authorization": f"Bearer {token}"},
            timeout=60
        )
        if resp.status_code != 200:
            raise Exception(f"下载备份失败: {resp.status_code}")

        backup_data = resp.json()
        cloud_words = set(w.lower() for w in backup_data.get('words', []))

        if merge:
            local_words = set(w.lower() for w in self._get_all_words())
            merged = local_words | cloud_words
            self._write_words(merged)
            return {
                "success": True,
                "action": "merged",
                "local_count": len(local_words),
                "cloud_count": len(cloud_words),
                "merged_count": len(merged),
                "new_words": len(merged - local_words),
                "version": backup_data.get('version', 0)
            }
        else:
            self._write_words(cloud_words)
            return {
                "success": True,
                "action": "replaced",
                "word_count": len(cloud_words),
                "version": backup_data.get('version', 0)
            }

    def _write_words(self, words: set):
        conn = sqlite3.connect(self.db_path)
        try:
            cursor = conn.cursor()
            cursor.execute('CREATE TABLE IF NOT EXISTS known_words (id INTEGER PRIMARY KEY AUTOINCREMENT, word TEXT UNIQUE NOT NULL, add_time TIMESTAMP DEFAULT CURRENT_TIMESTAMP)')
            cursor.execute('DELETE FROM known_words')
            for word in sorted(words):
                cursor.execute('INSERT OR IGNORE INTO known_words (word) VALUES (?)', (word,))
            conn.commit()
        finally:
            conn.close()

    def disconnect(self):
        self._tokens = None
        if os.path.exists(self.token_path):
            os.remove(self.token_path)

File: android/app/src/main/java/com/dreamword/app/onedrive/OneDriveBackup.kt
package com.dreamword.app.onedrive

import android.content.Context
import com.dreamword.app.data.KnownWordsDao
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.FormBody
import org.json.JSONArray
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.concurrent.TimeUnit

/**
 * OneDrive 备份/恢复（移植自 PC 版 src/modules/onedrive_backup.py）
 *
 * 使用 Microsoft Graph API + OAuth2 Device Code 流（无需公网回调，适合移动端）。
 * 词库备份到 OneDrive 的 App Root（special/approot），这是微软为应用提供的
 * 隔离文件夹，不会污染用户主目录。
 *
 * 与 PC 版的差异：
 *  - requests → OkHttp
 *  - token 缓存：明文 json 文件 → 应用私有 SharedPreferences（MODE_PRIVATE，
 *    其他应用无法读取；如需更强加密可换 androidx.security 的 EncryptedSharedPreferences）
 *  - SQLite 直接读写 → KnownWordsDao（已封装）
 *  - backup_meta.json 同步到云端（与 PC 一致），本地版本号也存 SharedPreferences
 *
 * Device Code 流说明：
 *  1. getDeviceCode() → 拿到 user_code + device_code + verification_url
 *  2. 用户在浏览器打开 verification_url，输入 user_code 完成授权
 *  3. 前端轮询 pollToken(device_code) → 授权完成后返回 token，否则返回 null
 */
class OneDriveBackup(
    private val context: Context,
    private val clientId: String,
    private val knownWords: KnownWordsDao
) {
    private val client = OkHttpClient.Builder()
        .connectTimeout(15, TimeUnit.SECONDS)
        .readTimeout(60, TimeUnit.SECONDS)
        .build()

    private val prefs by lazy {
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    // ============ Token 管理（加密存储）============

    private fun loadTokens(): JSONObject? {
        val json = prefs.getString(KEY_TOKENS, null) ?: return null
        return try { JSONObject(json) } catch (e: Exception) { null }
    }

    private fun saveTokens(tokens: JSONObject) {
        prefs.edit().putString(KEY_TOKENS, tokens.toString()).apply()
    }

    private fun clearTokens() {
        prefs.edit().remove(KEY_TOKENS).remove(KEY_LOCAL_VERSION).apply()
    }

    /**
     * 获取设备码（Device Code 流第一步）
     * 返回：{ user_code, device_code, verification_uri, expires_in, interval, message }
     */
    fun getDeviceCode(): JSONObject {
        val form = FormBody.Builder()
            .add("client_id", clientId)
            .add("scope", "Files.ReadWrite.AppFolder offline_access")
            .build()
        val req = Request.Builder()
            .url("https://login.microsoftonline.com/common/oauth2/v2.0/devicecode")
            .post(form)
            .build()
        client.newCall(req).execute().use { resp ->
            val body = resp.body?.string().orEmpty()
            if (!resp.isSuccessful) {
                throw Exception("获取设备码失败: ${resp.code} $body")
            }
            return JSONObject(body)
        }
    }

    /**
     * 轮询 token（Device Code 流第二步）
     * @return 授权完成返回 token JSON（已保存），未完成返回 null，错误抛异常
     */
    fun pollToken(deviceCode: String): JSONObject? {
        val form = FormBody.Builder()
            .add("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
            .add("client_id", clientId)
            .add("device_code", deviceCode)
            .build()
        val req = Request.Builder()
            .url("https://login.microsoftonline.com/common/oauth2/v2.0/token")
            .post(form)
            .build()
        client.newCall(req).execute().use { resp ->
            val body = resp.body?.string().orEmpty()
            val json = JSONObject(body)
            if (resp.isSuccessful) {
                // 记录过期时间戳（提前 60 秒）
                val expiresAt = System.currentTimeMillis() / 1000 +
                    json.optLong("expires_in", 3600) - 60
                json.put("expires_at", expiresAt)
                saveTokens(json)
                return json
            }
            // 授权中 / 慢一点：返回 null，让前端继续轮询
            val err = json.optString("error", "")
            if (err == "authorization_pending" || err == "slow_down") return null
            throw Exception("获取token失败: $body")
        }
    }

    private fun ensureToken(): String {
        val tokens = loadTokens() ?: throw Exception("未授权，请先完成 OneDrive 授权")
        val now = System.currentTimeMillis() / 1000
        if (now > tokens.optLong("expires_at", 0)) {
            refreshToken()
            return loadTokens()!!.optString("access_token")
        }
        return tokens.optString("access_token")
    }

    private fun refreshToken() {
        val tokens = loadTokens() ?: throw Exception("无 refresh_token，请重新授权")
        val rt = tokens.optString("refresh_token").orEmpty()
        if (rt.isEmpty()) throw Exception("无 refresh_token，请重新授权")
        val form = FormBody.Builder()
            .add("grant_type", "refresh_token")
            .add("client_id", clientId)
            .add("refresh_token", rt)
            .build()
        val req = Request.Builder()
            .url("https://login.microsoftonline.com/common/oauth2/v2.0/token")
            .post(form)
            .build()
        client.newCall(req).execute().use { resp ->
            val body = resp.body?.string().orEmpty()
            if (!resp.isSuccessful) throw Exception("刷新token失败: $body")
            val newTokens = JSONObject(body)
            newTokens.put("expires_at", System.currentTimeMillis() / 1000 +
                newTokens.optLong("expires_in", 3600) - 60)
            saveTokens(newTokens)
        }
    }

    // ============ Graph API 封装 ============

    private fun graphGet(path: String): JSONObject = graphRequest("GET", path, null)
    private fun graphPut(path: String, content: String): JSONObject =
        graphRequest("PUT", path, content)

    private fun graphRequest(method: String, path: String, content: String?): JSONObject {
        var token = ensureToken()
        var resp = doGraph(method, path, content, token)
        // 401 → 刷新 token 后重试一次
        if (resp.first == 401) {
            refreshToken()
            token = ensureToken()
            resp = doGraph(method, path, content, token)
        }
        val (code, body) = resp
        if (code !in 200..299) {
            throw Exception("Graph API 错误: $code $body")
        }
        return if (body.isNotBlank()) JSONObject(body) else JSONObject()
    }

    private fun doGraph(method: String, path: String, content: String?, token: String): Pair<Int, String> {
        val builder = Request.Builder()
            .url("$GRAPH_BASE$path")
            .header("Authorization", "Bearer $token")
        if (content != null) {
            builder.header("Content-Type", "application/json")
            builder.method(method, content.toRequestBody("application/json".toMediaType()))
        } else {
            builder.method(method, null)
        }
        client.newCall(builder.build()).execute().use { r ->
            return r.code to (r.body?.string().orEmpty())
        }
    }

    // ============ 业务方法 ============

    fun isAuthorized(): Boolean = try { ensureToken(); true } catch (e: Exception) { false }

    private fun getLocalVersion(): Int = prefs.getInt(KEY_LOCAL_VERSION, 0)
    private fun setLocalVersion(v: Int) = prefs.edit().putInt(KEY_LOCAL_VERSION, v).apply()

    /**
     * 备份已会词到 OneDrive（对应 Python backup）
     * @return { version, word_count, timestamp }
     */
    fun backup(): JSONObject {
        val words = knownWords.getAllWords()
        val newVersion = getLocalVersion() + 1
        val timestamp = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss", Locale.getDefault())
            .format(Date())

        val backupData = JSONObject()
            .put("version", newVersion)
            .put("timestamp", timestamp)
            .put("word_count", words.size)
            .put("words", JSONArray(words))

        // 触发 approot 创建（忽略错误）
        try { graphGet("/me/drive/special/approot") } catch (e: Exception) {}

        val filename = "known_words_v$newVersion.json"
        graphPut("/me/drive/special/approot:/$filename:/content", backupData.toString())

        // 写 meta（忽略错误，与 PC 版一致）
        try {
            val meta = JSONObject()
                .put("last_backup_version", newVersion)
                .put("last_backup_time", timestamp)
                .put("last_backup_count", words.size)
            graphPut("/me/drive/special/approot:/backup_meta.json:/content", meta.toString())
        } catch (e: Exception) {}

        setLocalVersion(newVersion)

        return JSONObject()
            .put("version", newVersion)
            .put("word_count", words.size)
            .put("timestamp", timestamp)
    }

    /**
     * 列出云端备份（对应 Python list_backups）
     * @return [{ name, size, last_modified }, ...] 按版本号倒序
     */
    fun listBackups(): JSONArray {
        val result = try {
            graphGet("/me/drive/special/approot/children")
        } catch (e: Exception) { return JSONArray() }

        val backups = mutableListOf<JSONObject>()
        val items = result.optJSONArray("value") ?: return JSONArray()
        for (i in 0 until items.length()) {
            val item = items.optJSONObject(i) ?: continue
            val name = item.optString("name")
            if (name.startsWith("known_words_v") && name.endsWith(".json")) {
                backups.add(JSONObject()
                    .put("name", name)
                    .put("size", item.optLong("size"))
                    .put("last_modified", item.optString("lastModifiedDateTime"))
                )
            }
        }
        // 按文件名（含版本号）倒序
        backups.sortByDescending { it.optString("name") }
        val out = JSONArray()
        backups.forEach { out.put(it) }
        return out
    }

    /**
     * 从 OneDrive 恢复（对应 Python restore）
     * @param backupName 指定备份文件名；null 则取 meta 里记录的最新版本
     * @param merge true=合并（并集），false=替换
     */
    fun restore(backupName: String?, merge: Boolean): JSONObject {
        val path = if (backupName != null) {
            "/me/drive/special/approot:/$backupName:/content"
        } else {
            val meta = graphGet("/me/drive/special/approot:/backup_meta.json:/content")
            val v = meta.optInt("last_backup_version", 0)
            if (v <= 0) throw Exception("云端无备份元数据")
            "/me/drive/special/approot:/known_words_v$v.json:/content"
        }

        // 下载备份内容（GET，可能返回大文件，单独走 ensureToken）
        val token = ensureToken()
        val req = Request.Builder()
            .url("$GRAPH_BASE$path")
            .header("Authorization", "Bearer $token")
            .get()
            .build()
        val backupData = client.newCall(req).execute().use { r ->
            if (!r.isSuccessful) throw Exception("下载备份失败: ${r.code}")
            JSONObject(r.body?.string().orEmpty())
        }

        val cloudWords = mutableSetOf<String>()
        val arr = backupData.optJSONArray("words")
        if (arr != null) for (i in 0 until arr.length()) {
            cloudWords.add(arr.getString(i).lowercase())
        }

        return if (merge) {
            val localWords = knownWords.getAllWords().map { it.lowercase() }.toMutableSet()
            val merged = (localWords + cloudWords).toList()
            knownWords.addWords(merged)
            JSONObject()
                .put("action", "merged")
                .put("local_count", localWords.size)
                .put("cloud_count", cloudWords.size)
                .put("merged_count", merged.size)
                .put("new_words", merged.size - localWords.size)
                .put("version", backupData.optInt("version", 0))
        } else {
            // 替换：先清空再写入（KnownWordsDao 无 clear，用删除全部+添加）
            for (w in knownWords.getAllWords()) knownWords.removeWord(w)
            knownWords.addWords(cloudWords.toList())
            JSONObject()
                .put("action", "replaced")
                .put("word_count", cloudWords.size)
                .put("version", backupData.optInt("version", 0))
        }
    }

    fun disconnect() = clearTokens()

    companion object {
        private const val PREFS_NAME = "onedrive_prefs"
        private const val KEY_TOKENS = "tokens"
        private const val KEY_LOCAL_VERSION = "last_backup_version"
        private const val GRAPH_BASE = "https://graph.microsoft.com/v1.0"
    }
}
